R/display.R

```diff
@@ -1,74 +1,171 @@
+#' 
+#' Allowed formats: png, pdf, svg
+#' 
+#' @examples
+#' \dontrun{
+#' display_images(
+#'   pdf = list(file = 'test.pdf'),
+#'   png = list(some.pngdata, width = 800, height = 600),
+#'   svg = '<svg></svg>')
+#' }
+#' 
 #' @export
-display_html = function(content) {
-    display1('text/html', content)
+display_images <- function(...) {
+    paramsets <- list(...)
+    
+    data <- list()
+    metadata <- namedlist()
+    for (format in names(paramsets)) {
+        params <- paramsets[[format]]
+        if (!is.list(params))  #`params` is just data
+            params <- list(params)
+        
+        args <- c(params, list(defer = TRUE))
+        def <- do.call(get_display_fun(format), args)
+        
+        data[[def$mimetype]] <- def$content
+        if (length(def$metadata) != 0)
+            metadata[[def$mimetype]] <- def$metadata
+    }
+    base_display(data, metadata)
 }
 
-#'Display PNG output
+#'Display HTML output
 #'
-#' Either data or filename must be passed.
-#' @param data The PNG data as a raw vector
-#' @param file The path to a PNG file or a connection
-#' @param width The width to display the image
-#' @param height The height to display the image
-#' @importFrom base64enc base64encode
+#' @param content  The HTML as a length 1 string vector
 #' @export
-display_png = function(data = NULL, file = NULL, width = NULL, height = NULL) {
+display_html <- function(content) {
+    display1('text/html', content)
+}
+
+display_binary_image <- function(mimetype, data = NULL, file = NULL, width = NULL, height = NULL, defer = FALSE) {
     if (!is.null(file)) {
         b64data <- base64encode(file)
     } else if (!is.null(data) && is.raw(data)) {
         b64data <- base64encode(data)
     } else stop('Either need to specify raw data as vector or filename/connection')
     
-    metadata = setNames(list(), character(0))
+    metadata = namedlist()
     if (!is.null(width))  metadata$width  <- width
     if (!is.null(height)) metadata$height <- height
-    display1('image/png', b64data, metadata)
+    
+    if (!defer)
+        display1(mimetype, b64data, metadata)
+    else
+        list(mimetype = mimetype, content = b64data, metadata = metadata)
+}
+
+#'Display PNG output
+#'
+#' Either data or file must be passed.
+#' @param data  The PNG data as a raw vector
+#' @param file  The path to a PNG file or a connection
+#' @param width  The width to display the image
+#' @param height  The height to display the image
+#' @param defer  Do not call \link{display1}, but return prepared data instead
+#' 
+#' @return If \code{defer} is TRUE, a list with the components:
+#' \itemize{
+#'  \item{mimetype}{\code{'image/png'}}
+#'  \item{content}{Base 64 encoded PNG data}
+#'  \item{metadata}{A list with \code{width} and \code{height} if given as parameters}
+#' }
+#' 
+#' @importFrom base64enc base64encode
+#' @export
+display_png <- function(data = NULL, file = NULL, width = NULL, height = NULL, defer = FALSE) {
```

Comment:
The `defer=FALSE` parameter doesn't feel like a good way to do things. If we sometimes need to prepare the data without sending it, there should be separate functions for prepare and send.